lib/include/aws_iot_defender.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2018 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file aws_iot_defender.h
+ * @brief
+ */
+
+#ifndef _AWS_IOT_DEFENDER_H_
+#define _AWS_IOT_DEFENDER_H_
+
+/* Standard includes. */
+#include <stdint.h>
+
+/**
+ * This enum type is the ID of metrics. It can be specified in AwsIotDefender_SetMetrics function.
+ */
+typedef enum
+{
+    AWS_IOT_DEFENDER_METRIC_TCP_ESTABLISHED_TOTAL,
+} AwsIotDefenderMetric_t;
+
+/**
+ * Return codes of defender functions.
+ */
+typedef enum
+{
+    AWS_IOT_DEFENDER_SUCCESS = 0,
+    AWS_IOT_DEFENDER_ALREADY_STARTED,
+    AWS_IOT_DEFENDER_PERIOD_TOO_SHORT,
+    AWS_IOT_DEFENDER_OUT_OF_MEMORY,
+
+    AWS_IOT_DEFENDER_INTERNAL_FAILURE
+} AwsIotDefenderError_t;
+
+/**
+ * Root cause code of failure, used in callback
+ */
+typedef enum
+{
+    AWS_IOT_DEFENDER_FAILURE_NONE = 0,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_CREATION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_CONNECTION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_SUBSCRIPTION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_PUBLISH,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_DISCONNECTION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_DELETION,
+    AWS_IOT_DEFENDER_FAILURE_METRICS_REJECTION
+} AwsIotDefenderFailureRootCause_t;
+
+/**
+ * Callback parameters.
+ */
+typedef struct AwsIotDefenderCallbackInfo
+{
+    const void * pPayload; /**< The received message if there is. */
```

Comment:
why void*?

Reply:
because MQTT defines AwsIotMqttPublishInfo_t:  const void * pPayload

Reply:
What format is the payload in?  If it's UTF-8 or binary, maybe make the pointer type char*/uint8_t*.

Reply:
I see. MQTT is more general, thus defining void *. For defender, uint8_t * should be the case. Good catch.